GitHubDataFetcher can't update a clonen repository if it contains modified files

Pulling updates from a repository in `GitHubDataFetcher` can fail with the following exception:

```
java.io.IOException: Could not pull to repository!
        at com.sap.sgs.phosphor.fosstars.data.github.LocalRepository.pull(LocalRepository.java:151)
        at com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher.loadLocalRepositoryFor(GitHubDataFetcher.java:287)
        at com.sap.sgs.phosphor.fosstars.data.github.GitHubDataFetcher.localRepositoryFor(GitHubDataFetcher.java:242)
        at com.sap.sgs.phosphor.fosstars.data.github.NumberOfCommits.fetchValueFor(NumberOfCommits.java:43)
        at com.sap.sgs.phosphor.fosstars.data.github.CachedSingleFeatureGitHubDataProvider.fetchValuesFor(CachedSingleFeatureGitHubDataProvider.java:28)
        at com.sap.sgs.phosphor.fosstars.data.github.CachedSingleFeatureGitHubDataProvider.fetchValuesFor(CachedSingleFeatureGitHubDataProvider.java:15)
        at com.sap.sgs.phosphor.fosstars.data.AbstractCachingDataProvider.doUpdate(AbstractCachingDataProvider.java:64)
        at com.sap.sgs.phosphor.fosstars.data.AbstractCachingDataProvider.doUpdate(AbstractCachingDataProvider.java:21)
        at com.sap.sgs.phosphor.fosstars.data.AbstractDataProvider.update(AbstractDataProvider.java:51)
        at com.sap.sgs.phosphor.fosstars.data.AbstractDataProvider.update(AbstractDataProvider.java:17)
        at com.sap.sgs.phosphor.fosstars.tool.github.SingleSecurityRatingCalculator.calculateFor(SingleSecurityRatingCalculator.java:83)
        at com.sap.sgs.phosphor.fosstars.tool.github.SingleSecurityRatingCalculator.calculateFor(SingleSecurityRatingCalculator.java:44)
        at com.sap.sgs.phosphor.fosstars.tool.github.MultipleSecurityRatingsCalculator.calculateFor(MultipleSecurityRatingsCalculator.java:73)
        at com.sap.sgs.phosphor.fosstars.tool.github.MultipleSecurityRatingsCalculator.calculateFor(MultipleSecurityRatingsCalculator.java:92)
        at com.sap.sgs.phosphor.fosstars.tool.github.MultipleSecurityRatingsCalculator.calculateFor(MultipleSecurityRatingsCalculator.java:15)
        at com.sap.sgs.phosphor.fosstars.tool.github.SecurityRatingCalculator.processConfig(SecurityRatingCalculator.java:249)
        at com.sap.sgs.phosphor.fosstars.tool.github.SecurityRatingCalculator.run(SecurityRatingCalculator.java:175)
        at com.sap.sgs.phosphor.fosstars.tool.github.SecurityRatingCalculator.main(SecurityRatingCalculator.java:94)
Caused by: org.eclipse.jgit.api.errors.CheckoutConflictException: Checkout conflict with files: 
src/Lucene.Net/overview.md
        at org.eclipse.jgit.api.MergeCommand.call(MergeCommand.java:401)
        at org.eclipse.jgit.api.PullCommand.call(PullCommand.java:367)
        at com.sap.sgs.phosphor.fosstars.data.github.LocalRepository.pull(LocalRepository.java:149)
        ... 17 more
Caused by: org.eclipse.jgit.errors.CheckoutConflictException: Checkout conflict with files: 
src/Lucene.Net/overview.md
        at org.eclipse.jgit.dircache.DirCacheCheckout.doCheckout(DirCacheCheckout.java:502)
        at org.eclipse.jgit.dircache.DirCacheCheckout.checkout(DirCacheCheckout.java:467)
        at org.eclipse.jgit.api.MergeCommand.call(MergeCommand.java:271)
        ... 19 more
```

It happens because for some reason the repository has modified files. The class should be able to overcome such issues. One possible way is to run `git reset --hard` before pulling. The class can also remove the corrupted repository if pulling still fails.

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java
@@ -269,29 +269,42 @@ public class GitHubDataFetcher {
         Files.delete(info.path());
       }
 
-      Optional<Repository> repository = openRepository(info.path());
-      if (!repository.isPresent()) {
+      try {
+        Optional<Repository> repository = openRepository(info.path());
+        if (!repository.isPresent()) {
+          Files.deleteIfExists(info.path());
+          clone(project, info.path());
+          repository = openRepository(info.path());
+        }
+
+        if (!repository.isPresent()) {
+          throw new IOException("Could not fetch project's repository!");
+        }
+
+        LocalRepository localRepository = new LocalRepository(info, repository.get());
+
+        if (shouldUpdate(localRepository)) {
+          LOGGER.info("Pulling updates from {} ...", project.url());
+          localRepository.reset();
+          localRepository.pull();
+        }
+
+        info.updated(Date.from(Instant.now()));
+        LOCAL_REPOSITORIES_INFO.put(project.url(), info);
+        return localRepository;
+      } catch (IOException e) {
+
+        // if something went wrong, then clean up info and cache,
+        // and remove the local repository if it exists
+        LOCAL_REPOSITORIES_INFO.remove(project.url());
+        LOCAL_REPOSITORIES.remove(project.url());
         Files.deleteIfExists(info.path());
-        clone(project, info.path());
-        repository = openRepository(info.path());
+
+        // then, re-throw the original exception
+        throw e;
+      } finally {
+        storeLocalRepositoriesInfo();
       }
-
-      if (!repository.isPresent()) {
-        throw new IOException("Could not fetch project's repository!");
-      }
-
-      LocalRepository localRepository = new LocalRepository(info, repository.get());
-
-      if (shouldUpdate(localRepository)) {
-        LOGGER.info("Pulling updates from {} ...", project.url());
-        localRepository.pull();
-      }
-
-      info.updated(Date.from(Instant.now()));
-      LOCAL_REPOSITORIES_INFO.put(project.url(), info);
-      storeLocalRepositoriesInfo();
-
-      return localRepository;
     }
   }
 

--- a/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/LocalRepository.java
+++ b/src/main/java/com/sap/sgs/phosphor/fosstars/data/github/LocalRepository.java
@@ -18,6 +18,7 @@ import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import org.apache.commons.io.IOUtils;
 import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.ResetCommand.ResetType;
 import org.eclipse.jgit.api.errors.GitAPIException;
 import org.eclipse.jgit.lib.Repository;
 import org.eclipse.jgit.revwalk.RevCommit;
@@ -136,6 +137,19 @@ public class LocalRepository implements AutoCloseable {
     }
 
     return Optional.of(commits.get(commits.size() - 1));
+  }
+
+  /**
+   * Resets the repository in a hard way by calling "git reset --hard".
+   *
+   * @throws IOException If something went wrong.
+   */
+  public void reset() throws IOException {
+    try (Git git = new Git(repository)) {
+      git.reset().setMode(ResetType.HARD).call();
+    } catch (GitAPIException e) {
+      throw new IOException("Could not reset the repository!", e);
+    }
   }
 
   /**

